### PR TITLE
Reset Password

### LINKED ISSUE #59: Forgot Password Functionality
**Description of the issue**
Users need to be able to reset their password if they forget by sending an email to reset their password.

**Files that need to be updated**
Will need to add some views to users.views and add a forgot password button to the frontend login page

**Link**
Will update when pr is created

### PR DESCRIPTION
**Pull request by courtmcgee2197**
*Please complete the checklist so the group can quickly understand your PR.*

### Category
- [x] **Feature**
- [ ] **Bug Fix**
- [ ] **Testing**
- [ ] **Documentation**
- [ ] **Other**

### Domain
- [ ] **Frontend**
- [ ] **Backend**
- [x] **Both**

### Description
This will allow users with an account to reset their password if they forgot their password.
### Related Issue(s)
This will solve issue #59 Forgot Password Functionality

### New Additions
No new files were added

### Modifications
Backend Files Altered:
backend/rebel_hubs/settings.py: Included an email protocol to send the reset password email. The email used is a dummy email I created for project purposes.
backend/users/serializers.py: Created a simple serializer for the email 
backend/users/serializers.py: I added all the urls needed for the password reset views
backend/users/views.py: I added the custom password reset view which would take in the email from the frontend and validate it.
Frontend Files Altered:
frontend/src/app/styles/register.css: Made minor changes to incorporate the new buttons as well as the unlv logo
frontend/src/components/Accounts/LoginForms.js: I added the forgot password button and register button. The forgot password button does not render a new page but shows a different view when clicked. The register button will route you to the register page.
frontend/src/components/Accounts/RegisterForms.js: Added the logo to this page

### New Dependencies
None

### Visuals
Screenshot (309)
Screenshot (308)
Screenshot (304)
Screenshot (305)
Screenshot (307)

### Architecture of Your Changes
The built in django system handles the resetting of the password using the django library views. Mostly just needed to connect the frontend to the views on the backend.

### Usage and Integration
This may be useful for the profile page with some alterations for changing your password if you are already logged in.

### Testing
In order test I created a user with an email I had access to. Attempted to send a password reset request. Once I received the email I changed the password and attempted to login with the new password.

### CODE REVIEW

**Comment by thecatfour** [comment]
When I click on the forgot password link, I get the following error message instead of the intended forgot password page:

image

However, when I change the users/urls.py file to delete the "template_name" field from the reset password urls, I get the following page:

image

I changed my password and logged in correctly, so it does correctly change the password for an account. Can anyone else recreate the issues that I am having, or is it just on my end?

**Comment by misherwa** [comment]
Hi Courtney, I am trying to register an account, but it seems that it's not working. Here's what it looks like in my end:
q2

I just want to make sure I understand the registration properly. When it comes to making an account, you mentioned using a dummy email. Do you recommend that I use a dummy email, or can I use my actual UNLV email? 

If I use a dummy email, does it have to be yahoo.com for the domain?

Also, I have to register again using the same email I used last time when I boot up the website again, right?

I do have some nitpicks. I think it looks a little weird having the "Forget password" and "Register" buttons so close together:
org


I notice that usually, websites will have a bit of space with their buttons, so I would also recommend this. For example, this is how the buttons or links look on the canvas login page:
 
cur

If you do take my nitpick into consideration, the buttons or links will look like this:
testcase

And this is what the section would look like:
q3


It's done by adding "style" to your buttons. You can set the "marginTop" to about 10 or 20 pixels to add spacing above your button or do "marginBottom" to add spacing below the button. You would have to do this in the LoginForms.js file found in components/Accounts.

This nitpick would also apply to your "Back" button on the Forgot Password page.

I know that my nitpicks are more towards the page than the actual reset password feature, but I'll try to be more helpful when I get the registration working.

File: frontend/src/components/Accounts/LoginForms.js
Comment (misherwa) [inline frontend/src/components/Accounts/LoginForms.js:117]: You can also add " style={{ marginTop: "5px" }}" for the button spacing for register. The line would look like "                    <button className="register" type="button" onClick={registerButtonClick} style={{ marginTop: "5px" }}>"

Or here's a picture for it:
q3

File: frontend/src/components/Accounts/LoginForms.js
Comment (misherwa) [inline frontend/src/components/Accounts/LoginForms.js:113]: You can add "style={{marginTop: "10px", marginBottom: "10px" }}" here for the button spacing. 

So the entire line would looks like " <button className="forgot-password" type="button" onClick={forgotPasswordButtonClick} style={{marginTop: "10px", marginBottom: "10px" }}>"

or here's a picture:

q2

**Comment by thecatfour** [comment]
The reset password function works for me now. However, I noticed a few issues.

### Issue 1
My first issue is during the reset password process. After resetting a password on the reset password form, I am redirected to this page: 

image

Clicking on the "Log in" link doesn't redirect me to the user login page and instead shows an error: 

image

Instead of being redirected to the frontend's login page, I seem to be getting redirected to a page on the backend.

### Issue 2
My second issue is when revisiting a reset password form. After I reset my password, clicking on the forgot password link again will result in the following page:

image

This is a small issue as I don't think people would try to click the link again after they already filled out the form, but we should keep this in mind for the future.

### Issue 3
Finally, it seems that the same email can be used multiple times to create multiple accounts. This probably won't be an issue when an authentication email is sent when creating an account, but I think this is important to at least point out. This multiple accounts for one email is shown by the forgot password function as shown by the following screenshot:

image


The reset password function works for me, but I think my first issue should be addressed in this PR. The reset password page currently redirect the user to a 404 page rather than the login page, which I think is a major issue. My second issue can probably be ignored for now, and the third issue should be its own PR.

**Comment by thecatfour** [comment]
If you want to do password resetting using the backend, then one solution to my first issue could be making a template html for the PasswordResetCompleteView. Here is an example of a basic redirect page that I made:

ezgif-5-c2d5f227c6

**Comment by thecatfour** [comment]
That is my main issue with this PR addressed. Now that I take a closer look at the email, I think it may not be a good idea to include the username in the forgot password email. If someone other than the user somehow got ahold of the email, it would let them log in as they would now know the username/password combination for the account. I think a message such as "Your RebelHub account associated with this email has issued a password reset. If this was you, then please use the link to choose a new password. If not, ignore this email." would be better than saying the username.

**Review by misherwa — APPROVED** [review]

**Review by leed85 — APPROVED** [review]

**Comment by leed85** [comment]
just need to fix the merge conflicts

**Review by thecatfour — APPROVED** [review]
Can be merged after merge conflicts are fixed

**Pull request merged**
